src/Poisson2DSparseMatrix.cpp

```diff
@@ -6,16 +6,191 @@
+        const real_matrix_type& sy = dg.sy();
+        const real_matrix_type& J  = dg.jacobian();
+        const real_matrix_type& Fscale = dg.fscale();
+
+        const index_vector_type& EToE = mshManager.get_EToE();
+        const index_vector_type& EToF = mshManager.get_EToF();
+
+        blitz::firstIndex ii;
+        blitz::secondIndex jj;
+        blitz::thirdIndex kk;
+
+        const index_type N = dg.order();
+
+        const index_vector_type& vmapM = dg.vmapM();
+        const index_vector_type& vmapP = dg.vmapP();
+
+        const index_matrix_type& Fmask = dg.fmask();
+
+        real_tensor3_type massEdge(Np, Np, Nfaces);
+        massEdge = 0.0;
+
+        Nodes1DProvisioner nodes1d(Nfp-1, 1, -1., 1.);
+
+        real_matrix_type V1D1(Nfp, Nfp), V1Dinv1(Nfp, Nfp);
+        real_matrix_type V1D2(Nfp, Nfp), V1Dinv2(Nfp, Nfp);
+        real_matrix_type V1D3(Nfp, Nfp), V1Dinv3(Nfp, Nfp);
+
+        real_matrix_type myTemp(Nfp, Nfp);
+
+        VandermondeBuilders vandermonde;
+
+        const real_vector_type& r = dg.r(), s = dg.s();
+        const real_matrix_type& nx = dg.nx(), ny  = dg.ny();
+
+        real_vector_type edge1(Nfp), edge2(Nfp), edge3(Nfp);
+        for (index_type i=0; i < Nfp; ++i)
+            edge1(i) = r(Fmask(i, 0));
+
+        vandermonde.computeVandermondeMatrix(edge1, V1D1, V1Dinv1);
+        myTemp = blitz::sum(V1Dinv1(kk,ii)*V1Dinv1(kk,jj), kk);
+
+        for (index_type i=0; i < Nfp; ++i) {
+            for (index_type j=0; j < Nfp; ++j) {
+                massEdge(Fmask(i, 0), Fmask(j, 0), 0) = myTemp(i, j);
+            }
+        }
+
+        for (index_type i=0; i < Nfp; ++i) {
+            edge2(i) = r(Fmask(i, 1));
+        }
+
+        vandermonde.computeVandermondeMatrix(edge2, V1D2, V1Dinv2);
+        myTemp = blitz::sum(V1Dinv2(kk,ii)*V1Dinv2(kk,jj), kk);
+        for (index_type i=0; i < Nfp; ++i) {
+            for (index_type j=0; j < Nfp; ++j) {
+                massEdge(Fmask(i, 1), Fmask(j, 1), 1) = myTemp(i, j);
+            }
+        }
+
+        for (index_type i=0; i < Nfp; ++i)
+            edge3(i) = s(Fmask(i, 2));
+
+        vandermonde.computeVandermondeMatrix(edge3, V1D3, V1Dinv3);
+        myTemp = blitz::sum(V1Dinv3(kk,ii)*V1Dinv3(kk,jj), kk);
+        for (index_type i=0; i < Nfp; ++i) {
+            for (index_type j=0; j < Nfp; ++j) {
+                massEdge(Fmask(i, 2), Fmask(j, 2), 2) = myTemp(i, j);
+            }
+        }
+
+		real_vector_type ubcCol(Nfp*Nfaces*K), qbcCol(Np*Nfaces*K);
+		ubcCol = 0.0;
+		qbcCol = 0.0;
+		reshapeMatTo1D(ubc, ubcCol.data(), false);
+		reshapeMatTo1D(qbc, qbcCol.data(), false);
+
+        for (index_type k=0; k < K; ++k) {
+            const index_type k1 = k;
+
+            real_matrix_type Dx(Np, Np), Dy(Np, Np);
+
+            Dx = rx(0, k)*Dr + sx(0, k)*Ds;
+            Dy = ry(0, k)*Dr + sy(0, k)*Ds;
+
+            for (index_type f1=0; f1 < Nfaces; ++f1) {
+                index_type k2 = EToE(Nfaces*k1 + f1);
```

Comment:
You index `EToE` and `EToF` as though they are matrices, yet you store them as arrays instead of an `index_matrix_type`. Just wondering why? The indexing would be more natural stored as a matrix.

Reply:
iirc, this decision was made in case we wanted to use one MeshManager to handle both triangles and quads (or say, a mesh containing both). Then, we could keep the data structures the same and just leaf through the array with some kind of sparse index array. Maybe it's something we'll never use though, so it could be worth refactoring these to matrices.

Reply:
:+1: